src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -697,10 +705,11 @@ private static EntitySourceType GetEntitySourceType(string entityName, Entity en
         /// <param name="entity"></param>
         /// <param name="databaseTable"></param>
         /// <exception cref="InvalidOperationException"></exception>
-        private void AddForeignKeysForRelationships(
+        private void ProcessRelationships(
             string entityName,
             Entity entity,
-            DatabaseTable databaseTable)
+            DatabaseTable databaseTable,
+            Dictionary<string, DatabaseObject> sourceObjects)
```

Comment:
1. EntityToDatabaseObject stores the mapping from EntityName to DatabaseObject.
2. sourceObjects stores the mapping from source name to DatabaseObject. Multiple entities can have different entity names yet the same backing source table in the database. Keeping track of all the source objects help us avoid creating database objects for linking entities again in case the linking entities are backed by a source for which we have already inferred the database object.